ui-modules/blueprint-composer/app/components/providers/recently-used-service.provider.js

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+export function recentlyUsedServiceProvider() {
+    return {
+        $get: ['$log', function ($log) { return new RecentlyUsedService($log); }]
+    }
+}
+
+export const PREFIX = 'org.apache.brooklyn.composer.lastUsed:';
+
+function RecentlyUsedService($log) {
+    let service = {};
+    
+    service.getId = (brooklynObject) => {
+        if (typeof brooklynObject === 'string') return brooklynObject;
+        return (brooklynObject.containingBundle || '?') + "::" + 
+            (brooklynObject.symbolicName || '?') + ":" + (brooklynObject.version || '?');
+    };
+    
+    service.markUsed = (item, when) => {
+        let id = service.getId(item);
+        if (when) {
+            let old = service.getLastUsed(id);
+            if (old > when) return;
+        } else {
+            when = Date.now();
+        }
+        sessionStorage.setItem(PREFIX+id, when);
```

Comment:
Storing each item in the `sessionStorage` can lead to crippled instances as there is a limited number of items you can store them. Instead, a good solution is to store the entire `Map` of items as a `JSON.stringify()` into one `sessionStorage.setItem()`

Reply:
hmm, i thought stringifying a potentially very large map on every change would be less performant than using session storage.  do you have more info on the limits?  (the limits i've seen are on _size_ not number of entries.)

Reply:
We did hit the maximum size on the earlier day of the composer while saving the generated icons. Now, I think it's the actual size of the `localStorage` (5MB if I recall correctly) but don't remember if there is a limit per item or in total.

Nonetheless, we need to handle this because `sessionStorage.setItem()` will throw if there is no more space left.

Reply:
seems like the issue is that it can through; will handle that with `try ... catch`, good point.  won't convert to stringified map unless you know a good reason to do that.